Sources/SourceKitBazelBSP/RequestHandlers/BuildTargets/BazelQueryParser.swift

```diff
@@ -17,114 +17,196 @@
+        from targets: [BlazeQuery_Target],
         supportedRuleTypes: Set<String>,
         rootUri: String,
         toolchainPath: String
     ) throws -> [(BuildTarget, [URI])] {
+        var result: [(BuildTarget, [URI])] = []
+        let srcMap = buildSourceFilesMap(targets)
+
+        for target in targets {
+            // make sure BlazeQuery_Target.Discriminator is rule type
+            guard target.type == .rule else {
+                continue
+            }
+
+            let rule = target.rule
+
+            let id: URI = try rule.name.toTargetId(rootUri: rootUri)
+
+            let baseDirectory: URI = try rule.name.toBaseDirectory(rootUri: rootUri)
+
+            // get test_only
+            let testOnly = rule.attribute.first { $0.name == "testonly" }?.booleanValue ?? false
 
-        // FIXME: Most of this logic is hacked together and not thought through, with the
-        // sole intention of getting the example project to work.
-        // Need to understand what exactly we can receive from the queries to know how to properly
-        // parse this info.
-
-        var targets: [(BuildTarget, [URI])] = []
-        for child in (xml.children ?? []) {
-            if child.name != "rule" { continue }
-            guard let childElement = child as? XMLElement else { continue }
-            let className = childElement.attribute(forName: "class")?.stringValue ?? ""
-            guard supportedRuleTypes.contains(className) else { continue }
-            if let data = try getTargetForLibrary(childElement, className, rootUri, toolchainPath) {
-                targets.append(data)
+            // BuildTargetCapabilities
+            let capabilities = BuildTargetCapabilities(
+                canCompile: true,
+                canTest: testOnly,
+                canRun: false,
+                canDebug: false
+            )
+
+            // get language
+            let isSwift = target.rule.ruleClass.contains("swift")
+
+            // get direct upstream dependencies
+            let deps: [BuildTargetIdentifier] = try rule.attribute.flatMap { attr in
+                attr.name == "deps" ? attr.stringListValue : []
+            }.map { label in
+                let id = try label.toTargetId(rootUri: rootUri)
+                return BuildTargetIdentifier(uri: id)
+            }
+
+            // get srcs
+            let srcs: [URI] = try rule.attribute.flatMap { attr in
+                attr.name == "srcs" ? attr.stringListValue : []
+            }.compactMap { src in
+                guard let path = srcMap[src] else {
+                    let error = BazelTargetParserError.noSrcFound(src)
+                    logger.info("\(error, privacy: .public)")
+                    return nil
+                }
+                return try URI(string: path)
             }
+
+            let data = SourceKitBuildTarget(
+                toolchain: try URI(string: "file://" + toolchainPath)
+            ).encodeToLSPAny()
```

Comment:
Since this is the same for all libraries, we can move it outside of the loop for a small perf gain.